compiler/vm/vm.nim

```diff
@@ -267,71 +432,104 @@ proc copyValue(src: PNode): PNode =
+
+    x = TFullReg(kind: rkValue, handle: loc)
+  of rkNimNode:
+    x = TFullReg(kind: rkNimNode, nimNode: copyValue(y.nimNode))
+
+  of rkRegAddr:
+    x = TFullReg(kind: rkRegAddr, regAddr: y.regAddr)
+
+
+proc asgnRef(x: var TFullReg, y: TFullReg, mm: var VmMemoryManager) =
+  if x.kind == rkValue:
+    mm.resetLocation(x.handle.byteView(), x.handle.typ)
+
+  case y.kind
+  of rkNone, rkInt, rkFloat: assert false
+  of rkAddress: x = TFullReg(kind: rkAddress, addrVal: y.addrVal, addrTyp: y.addrTyp, isValid: y.isValid)
+  of rkNimNode: x = TFullReg(kind: rkNimNode, nimNode: y.nimNode)
+  of rkHandle, rkValue:
+    assert y.handle.typ.kind == akRef
+    let loc = mm.allocSingleLocation(y.handle.typ)
+    copyToLocation(dest = loc, src = y.handle, mm, reset = false)
+
+    x = TFullReg(kind: rkValue, handle: loc)
+  of rkRegAddr:
+    x = TFullReg(kind: rkRegAddr, regAddr: y.regAddr)
+
+
 
 proc fastAsgnComplex(x: var TFullReg, y: TFullReg) =
-  x.ensureKind(y.kind)
-  case x.kind
-  of rkNone: discard
-  of rkInt: x.intVal = y.intVal
-  of rkFloat: x.floatVal = y.floatVal
-  of rkNode: x.node = y.node
-  of rkRegisterAddr: x.regAddr = y.regAddr
-  of rkNodeAddr: x.nodeAddr = y.nodeAddr
+  case y.kind
+  of rkNone: x.reset()
+  of rkInt, rkFloat, rkAddress: copyMem(addr x, unsafeAddr y, sizeof(TFullReg))
+  of rkHandle, rkValue: x = TFullReg(kind: rkHandle, handle: y.handle)
+  of rkNimNode: x = TFullReg(kind: rkNimNode, nimNode: y.nimNode)
+
+  of rkRegAddr: x = TFullReg(kind: rkRegAddr, regAddr: y.regAddr)
+
 
-proc writeField(n: var PNode, x: TFullReg) =
+template toPointer(x: TypedHandle): untyped =
+  x.h.toPointer
+
+proc writeField(h: TypedHandle, x: TFullReg, mm: var VmMemoryManager) =
   case x.kind
-  of rkNone: discard
+  of rkNone: assert false # This hints at a programming error somewhere
   of rkInt:
-    if n.kind == nkNilLit:
-      n[] = TNode(kind: nkIntLit) # ideally, `nkPtrLit`
-    n.intVal = x.intVal
-  of rkFloat: n.floatVal = x.floatVal
-  of rkNode: n = copyValue(x.node)
-  of rkRegisterAddr: writeField(n, x.regAddr[])
-  of rkNodeAddr: n = x.nodeAddr[]
-
-proc putIntoReg(dest: var TFullReg; n: PNode) =
-  case n.kind
-  of nkStrLit..nkTripleStrLit:
-    dest = TFullReg(kind: rkNode, node: newStrNode(nkStrLit, n.strVal))
-  of nkIntLit: # use `nkPtrLit` once this is added
-    if dest.kind == rkNode: dest.node = n
-    elif n.typ != nil and n.typ.kind in PtrLikeKinds:
-      dest = TFullReg(kind: rkNode, node: n)
-    else:
-      dest = TFullReg(kind: rkInt, intVal: n.intVal)
-  of {nkCharLit..nkUInt64Lit} - {nkIntLit}:
-    dest = TFullReg(kind: rkInt, intVal: n.intVal)
-  of nkFloatLit..nkFloat128Lit:
-    dest = TFullReg(kind: rkFloat, floatVal: n.floatVal)
-  else:
-    dest = TFullReg(kind: rkNode, node: n)
-
-proc regToNode(x: TFullReg): PNode =
+    assert h.typ.kind == akInt
+    deref(h).intVal = x.intVal
+  of rkFloat:
+    assert h.typ.kind == akFloat
```

Comment:
I'd recommend avoiding assertions where possible.

Instead, put absolute trust in the codegen, and debug troublesome places where things acts weird.

Assertions in VM are unfortunately useless, as you have no trace of the instruction stream that leads to the erroneous state.

Also, accessing wrong discriminator field is checked (in debug builds), so there is much less value in manually checking too.

Reply:
I rapidly changed large amounts of code multiple times and assertions helped me to make sure that preconditions we're not violated after changes (which could also hint at obsolete code) and to catch bugs early. I'm also using them as tool for documentation.

In the error handling overhaul that I'm planning to do in separate PR prior to merging this one, I'd have replaced assertions in the VM with something like `vmAssert`, which, in debug builds, would throw an error which would be caught by a `try/except` around `rawExecute` which would then print a guest-code stack-trace and either do a `globalReport` or turn the error into an `nkError`.

In the code you commented on, the assertion is meant for the expression on the left, where no variant field access happens.

If okay, I'd keep the assertion till I finish this PR, but would remove/replace them prior to the merge, if wanted

Reply:
If you find them useful for debugging, please keep them. I'd prefer a design where we have as little assertions as humanly possible, but that can wait until things cleared up.